### DISCUSSION TITLE
Advice on a use case

### DISCUSSION THREAD

**Discussion by PeterGillbr** [bellrichm/WeeWX-MQTTSubscribe#132] [Q&A] [answered]
It would be highly appreciated if someone would have the time to provide some advice on parametrization of  WeeWX-MQTTSubscribe provided it is applicable to the use case below, which I assume.
The use case:
Data from a weather station is captured by Weewx installed on a Raspberry PI (Python 3) via the FineOffsetUSB interface.
An external extra temp sensor (Watertemp) is providing data to the R-Pi through MQTT packages. The MQTT topic looks like: “Jettystation/Watertemp” and the payload is a string (e.g 19.1)
A mosquitto server is installed and is running on the R-Pi. 
All of the above is working.
The target is to extend the standard weewx database (weewx.sdb) with a Watertemp column, updated in the same manor weewx is updating the data base.
So:
Should WeeWX-MQTTSubscribe be installed as a driver or service?
How should the configuration parameters be set?
I realize this is a pretty comprehensive question, but any advice would be highly appreciated.
Thanks in advance

**Comment by bellrichm** [comment, marked as answer]
Here is a rough guide. It does assume some WeeWX and MQTT knowledge.
1) Determine whether to run as a driver or service.
If you are augmenting an existing driver's data, run MQTTSubscribe as a service. Otherwise, run it as a driver.
It sounds like you have a WeeWX driver pulling data from the FineOffsetUSB interface, so you want to run as a service to add additional observations.
2) Determine the payload 'type'; individual, json, keyword.
From your comment, your 'type' appears to be individual.
Here is an example configuration for that type, https://raw.githubusercontent.com/wiki/bellrichm/WeeWX-MQTTSubscribe/individual-example.md
3) Determine what data transformation/conversion needs to be done on the incoming data.
You want the name to be 'Watertemp'
What units is the incoming data, Celsius or Farenheight?  I'm guessing Celsius.
Does it contain a 'total' value (for example rain total), etc... It looks like a 'no' to this.
4) Determine the MQTT parameters
Host, port, username, etc.
5) Since you want to use a new observation, Watertemp (versus using an existing one like extraTemp1), you are going to need to understand how to customize WeeWX, https://weewx.com/docs/customizing.htm
a) You will need to update the database, https://weewx.com/docs/customizing.htm#archive_database
b) You will also need to understand 'units' and 'unit_groups', https://weewx.com/docs/customizing.htm#customizing_units This talks about writing code, but MQTTSubscribe has some options to configure it. See, https://github.com/bellrichm/WeeWX-MQTTSubscribe/wiki/Customizing-units-and-unit-groups
c) Then you will want to read up on customizing reports, etc.

So, putting it all together you will end up with something like this
```
[MQTTSubscribeService]
    host = 
    port = 

    [[weewx]]
        [[[observations]]]
            Watertemp = group_temperature

    [[message_callback]]
        type = individual

    [[topics]]
        unit_system = METRIC

        [[[Jettystation/Watertemp]]]
            name = Watertemp
```

Note, MQTTSubscribeService has some 'date/time quality control' checks. So, we might have to do some configuration of those. We can look into that once you have it up and running. The detail can be found here, https://github.com/bellrichm/WeeWX-MQTTSubscribe/wiki/Date-Time-processing

**Reply by PeterGillbr** [reply]
Thanks so much for this answer and the time you spent writing it! Highly, highly appreciated! Now everything works exactly how it is supposed! Thanks again!

**Reply by bellrichm** [reply]
Glad you have it working. Because there are so many different possible inputs, I have struggled with writing a 'recipe'. But, you have me thinking that I could write some FAQs.

**Reply by earlthesquirrel** [reply]
FYI -- I had some devices that posted to "different" topics that I wanted to map to existing Weemx values.  It is fairly simple to write the code that will listen for those values and then re-post them to the existing Weemx value topics.  (In a couple cases I could update the configuration of the devices to change their topic values, but in some cases I could not).  I can share this code if anybody is interested.
